# Formula 1 World Championship Analysis
## Exploratory Data Analysis
* **Goal:** TBA

In [42]:
# Import libraries for EDA
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [43]:
# Create function for safe loading of csv files
def load_csv(filepath):
    try:
        df = pd.read_csv(filepath)
        print(f'{filepath} successfully loaded')
        return df
    except FileNotFoundError:
        print(f'{filepath} not found')
    except Exception as e:
        print(f'Error loading {filepath}: {e}')

In [44]:
# Load CSV Files

circuits = load_csv('../data/circuits.csv')
constructor_results = load_csv('../data/constructor_results.csv')
constructor_standings = load_csv('../data/constructor_standings.csv')
constructors = load_csv('../data/constructors.csv')
driver_standings = load_csv('../data/driver_standings.csv')
drivers = load_csv('../data/drivers.csv')
lap_times = load_csv('../data/lap_times.csv')
pit_stops = load_csv('../data/pit_stops.csv')
qualifying = load_csv('../data/qualifying.csv')
races = load_csv('../data/races.csv')
results = load_csv('../data/results.csv')
seasons = ('../data/seasons.csv')
sprint_results = load_csv('../data/sprint_results.csv')
status = load_csv('../data/status.csv')

../data/circuits.csv successfully loaded
../data/constructor_results.csv successfully loaded
../data/constructor_standings.csv successfully loaded
../data/constructors.csv successfully loaded
../data/driver_standings.csv successfully loaded
../data/drivers.csv successfully loaded
../data/lap_times.csv successfully loaded
../data/pit_stops.csv successfully loaded
../data/qualifying.csv successfully loaded
../data/races.csv successfully loaded
../data/results.csv successfully loaded
../data/sprint_results.csv successfully loaded
../data/status.csv successfully loaded


In [45]:
# Create function for inspection of dataset
def dataset_inspection(df):
    print('Dataset Shape:')
    print(f'Rows: {df.shape[0]} -- Columns: {df.shape[1]}')

    print('\nDataset Information:')
    df.info()

    print('\nMissing Values:')
    display(df.isna().sum())

    print('\nDatset Head:')
    display(df.head())
    
    return df

In [46]:
dataset_inspection(drivers)

Dataset Shape:
Rows: 861 -- Columns: 9

Dataset Information:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 861 entries, 0 to 860
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   driverId     861 non-null    int64 
 1   driverRef    861 non-null    object
 2   number       861 non-null    object
 3   code         861 non-null    object
 4   forename     861 non-null    object
 5   surname      861 non-null    object
 6   dob          861 non-null    object
 7   nationality  861 non-null    object
 8   url          861 non-null    object
dtypes: int64(1), object(8)
memory usage: 60.7+ KB

Missing Values:


driverId       0
driverRef      0
number         0
code           0
forename       0
surname        0
dob            0
nationality    0
url            0
dtype: int64


Datset Head:


,driverId,driverRef,number,code,forename,surname,dob,nationality,url
0,1,hamilton,44,HAM,Lewis,Hamilton,1985-01-07,British,http://en.wikipedia.org/wiki/Lewis_Hamilton
1,2,heidfeld,\N,HEI,Nick,Heidfeld,1977-05-10,German,http://en.wikipedia.org/wiki/Nick_Heidfeld
2,3,rosberg,6,ROS,Nico,Rosberg,1985-06-27,German,http://en.wikipedia.org/wiki/Nico_Rosberg
3,4,alonso,14,ALO,Fernando,Alonso,1981-07-29,Spanish,http://en.wikipedia.org/wiki/Fernando_Alonso
4,5,kovalainen,\N,KOV,Heikki,Kovalainen,1981-10-19,Finnish,http://en.wikipedia.org/wiki/Heikki_Kovalainen


,driverId,driverRef,number,code,forename,surname,dob,nationality,url
0,1,hamilton,44,HAM,Lewis,Hamilton,1985-01-07,British,http://en.wikipedia.org/wiki/Lewis_Hamilton
1,2,heidfeld,\N,HEI,Nick,Heidfeld,1977-05-10,German,http://en.wikipedia.org/wiki/Nick_Heidfeld
2,3,rosberg,6,ROS,Nico,Rosberg,1985-06-27,German,http://en.wikipedia.org/wiki/Nico_Rosberg
3,4,alonso,14,ALO,Fernando,Alonso,1981-07-29,Spanish,http://en.wikipedia.org/wiki/Fernando_Alonso
4,5,kovalainen,\N,KOV,Heikki,Kovalainen,1981-10-19,Finnish,http://en.wikipedia.org/wiki/Heikki_Kovalainen
...,...,...,...,...,...,...,...,...,...
856,858,sargeant,2,SAR,Logan,Sargeant,2000-12-31,American,http://en.wikipedia.org/wiki/Logan_Sargeant
857,859,lawson,30,LAW,Liam,Lawson,2002-02-11,New Zealander,http://en.wikipedia.org/wiki/Liam_Lawson
858,860,bearman,38,BEA,Oliver,Bearman,2005-05-08,British,http://en.wikipedia.org/wiki/Oliver_Bearman
859,861,colapinto,43,COL,Franco,Colapinto,2003-05-27,Argentinian,http://en.wikipedia.org/wiki/Franco_Colapinto


In [47]:
dataset_inspection(results)

Dataset Shape:
Rows: 26759 -- Columns: 18

Dataset Information:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26759 entries, 0 to 26758
Data columns (total 18 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   resultId         26759 non-null  int64  
 1   raceId           26759 non-null  int64  
 2   driverId         26759 non-null  int64  
 3   constructorId    26759 non-null  int64  
 4   number           26759 non-null  object 
 5   grid             26759 non-null  int64  
 6   position         26759 non-null  object 
 7   positionText     26759 non-null  object 
 8   positionOrder    26759 non-null  int64  
 9   points           26759 non-null  float64
 10  laps             26759 non-null  int64  
 11  time             26759 non-null  object 
 12  milliseconds     26759 non-null  object 
 13  fastestLap       26759 non-null  object 
 14  rank             26759 non-null  object 
 15  fastestLapTime   26759 non-null  object 

resultId           0
raceId             0
driverId           0
constructorId      0
number             0
grid               0
position           0
positionText       0
positionOrder      0
points             0
laps               0
time               0
milliseconds       0
fastestLap         0
rank               0
fastestLapTime     0
fastestLapSpeed    0
statusId           0
dtype: int64


Datset Head:


,resultId,raceId,driverId,constructorId,number,grid,position,positionText,positionOrder,points,laps,time,milliseconds,fastestLap,rank,fastestLapTime,fastestLapSpeed,statusId
0,1,18,1,1,22,1,1,1,1,10.0,58,1:34:50.616,5690616,39,2,1:27.452,218.300,1
1,2,18,2,2,3,5,2,2,2,8.0,58,+5.478,5696094,41,3,1:27.739,217.586,1
2,3,18,3,3,7,7,3,3,3,6.0,58,+8.163,5698779,41,5,1:28.090,216.719,1
3,4,18,4,4,5,11,4,4,4,5.0,58,+17.181,5707797,58,7,1:28.603,215.464,1
4,5,18,5,1,23,3,5,5,5,4.0,58,+18.014,5708630,43,1,1:27.418,218.385,1


,resultId,raceId,driverId,constructorId,number,grid,position,positionText,positionOrder,points,laps,time,milliseconds,fastestLap,rank,fastestLapTime,fastestLapSpeed,statusId
0,1,18,1,1,22,1,1,1,1,10.0,58,1:34:50.616,5690616,39,2,1:27.452,218.300,1
1,2,18,2,2,3,5,2,2,2,8.0,58,+5.478,5696094,41,3,1:27.739,217.586,1
2,3,18,3,3,7,7,3,3,3,6.0,58,+8.163,5698779,41,5,1:28.090,216.719,1
3,4,18,4,4,5,11,4,4,4,5.0,58,+17.181,5707797,58,7,1:28.603,215.464,1
4,5,18,5,1,23,3,5,5,5,4.0,58,+18.014,5708630,43,1,1:27.418,218.385,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
26754,26760,1144,825,210,20,14,16,16,16,0.0,57,\N,\N,57,1,1:25.637,222.002,11
26755,26761,1144,859,215,30,12,17,17,17,0.0,55,\N,\N,52,12,1:28.751,214.212,5
26756,26762,1144,822,15,77,9,\N,R,18,0.0,30,\N,\N,14,19,1:29.482,212.462,130
26757,26763,1144,861,3,43,20,\N,R,19,0.0,26,\N,\N,5,17,1:29.411,212.631,5


In [48]:
# Merge both CSV 'drivers' and 'result'
drivers_results_df = results.merge(drivers, on='driverId')

# Display merged DataFrame
display(drivers_results_df.head())

,resultId,raceId,driverId,constructorId,number_x,grid,position,positionText,positionOrder,points,...,fastestLapSpeed,statusId,driverRef,number_y,code,forename,surname,dob,nationality,url
0,1,18,1,1,22,1,1,1,1,10.0,...,218.300,1,hamilton,44,HAM,Lewis,Hamilton,1985-01-07,British,http://en.wikipedia.org/wiki/Lewis_Hamilton
1,2,18,2,2,3,5,2,2,2,8.0,...,217.586,1,heidfeld,\N,HEI,Nick,Heidfeld,1977-05-10,German,http://en.wikipedia.org/wiki/Nick_Heidfeld
2,3,18,3,3,7,7,3,3,3,6.0,...,216.719,1,rosberg,6,ROS,Nico,Rosberg,1985-06-27,German,http://en.wikipedia.org/wiki/Nico_Rosberg
3,4,18,4,4,5,11,4,4,4,5.0,...,215.464,1,alonso,14,ALO,Fernando,Alonso,1981-07-29,Spanish,http://en.wikipedia.org/wiki/Fernando_Alonso
4,5,18,5,1,23,3,5,5,5,4.0,...,218.385,1,kovalainen,\N,KOV,Heikki,Kovalainen,1981-10-19,Finnish,http://en.wikipedia.org/wiki/Heikki_Kovalainen


In [49]:
# Identify drivers with the most race wins
winners_count = drivers_results_df[drivers_results_df['positionOrder'] == 1]
print(winners_count.groupby(['surname','forename']).size().sort_values(ascending=False).head(20))

surname     forename 
Hamilton    Lewis        105
Schumacher  Michael       91
Verstappen  Max           63
Vettel      Sebastian     53
Prost       Alain         51
Senna       Ayrton        41
Alonso      Fernando      32
Mansell     Nigel         31
Stewart     Jackie        27
Clark       Jim           25
Lauda       Niki          25
Fangio      Juan          24
Rosberg     Nico          23
Piquet      Nelson        23
Hill        Damon         22
Räikkönen   Kimi          21
Häkkinen    Mika          20
Moss        Stirling      16
Button      Jenson        15
Fittipaldi  Emerson       14
dtype: int64


In [50]:
# Calculate the total World Championship points by driver
print(drivers_results_df.groupby(['surname', 'forename'])['points'].sum().sort_values(ascending=False).head(20))

surname      forename 
Hamilton     Lewis        4820.5
Vettel       Sebastian    3098.0
Verstappen   Max          2912.5
Alonso       Fernando     2329.0
Räikkönen    Kimi         1873.0
Bottas       Valtteri     1788.0
Rosberg      Nico         1594.5
Pérez        Sergio       1585.0
Schumacher   Michael      1566.0
Leclerc      Charles      1363.0
Ricciardo    Daniel       1320.0
Button       Jenson       1235.0
Sainz        Carlos       1203.5
Massa        Felipe       1167.0
Webber       Mark         1047.5
Norris       Lando         950.0
Prost        Alain         798.5
Russell      George        664.0
Barrichello  Rubens        658.0
Senna        Ayrton        614.0
Name: points, dtype: float64


**Note:**

Comparing the number of race wins with the total World Championship (WM) points reveals some interesting differences. Lewis Hamilton ranks **first** in both categories. Michael Schumacher is **second** in terms of race wins, but only **ninth** when ranked by total WM points.

Hamilton has **14 more race wins** than Schumacher. This gap becomes even more noticeable when comparing their total World Championship points, where Hamilton has **[insert point difference]** more points than Schumacher.

In [51]:
# Check for winners count by nationality
nat_winners_count = drivers_results_df[drivers_results_df['positionOrder'] == 1]
print(nat_winners_count.groupby('nationality').size().sort_values(ascending=False).head(20))

nationality
British          317
German           179
Brazilian        101
French            81
Dutch             63
Finnish           57
Australian        45
Italian           43
Austrian          41
Argentine         38
Spanish           36
American          33
Canadian          17
New Zealander     12
Swedish           12
Belgian           11
South African     10
Mexican            8
Monegasque         8
Colombian          7
dtype: int64


**Note:**

The analysis shows that British drivers have achieved the highest number of race wins (**317**), followed by German (**179**) and Brazilian (**101**) drivers. Most of the top-ranked nationalities are European, indicating that European countries have produced many of Formula 1's most successful drivers.

In [52]:
# Check Metrics of grid column
drivers_results_df['grid'].describe()

count    26759.000000
mean        11.134796
std          7.202860
min          0.000000
25%          5.000000
50%         11.000000
75%         17.000000
max         34.000000
Name: grid, dtype: float64

In [53]:
# Check how often does driver win in pole positions
pole_winners = drivers_results_df[(drivers_results_df['grid'] == 1) & (drivers_results_df['positionOrder'] == 1)]
print(len(pole_winners))

481


In [54]:
# Check for winning rate
pole_starts = drivers_results_df[drivers_results_df['grid'] == 1]
pole_win_rate = (pole_starts['positionOrder'] == 1).mean()
print(f'{pole_win_rate*100:.2f}%')

42.34%


**Note:**

The analysis shows that starting a race from pole position provides a clear advantage, but does not guarantee a race win. Out of all races in which a driver started from pole position, **481 races were won by the pole sitter**, resulting in a **win rate of 42.34%**.

This means that in more than half of the cases (**57.66%**), the driver starting from pole position did not win the race. Therefore, while pole position appears to provide an advantage, other factors also influence the final race result.

In [55]:
# Create new column to check how much positions are gained
drivers_results_df['positions_gained'] = (drivers_results_df['grid'] - drivers_results_df['positionOrder'])

drivers_results_df['positions_gained'].describe()

count    26759.000000
mean        -1.659255
std          9.613308
min        -39.000000
25%         -5.000000
50%          0.000000
75%          4.000000
max         30.000000
Name: positions_gained, dtype: float64

**Note:**

The positions_gained analysis shows that drivers changed their starting position by an average of -1.66 positions during a race. The median is 0, indicating that the typical result was no change in position. The values range from a maximum loss of 39 positions to a maximum gain of 30 positions. Overall, the results show that while drivers can gain a significant number of positions during a race, position losses are also common.

In [56]:
# Cut 'position_gained' into 3 labels with pd.cut
position_change = pd.cut(drivers_results_df['positions_gained'],
                         bins=[-float('inf'), -1, 0, float('inf')],
                         labels=['Lost Position', 'No Change', 'Gained Position']
                         )
print(position_change.value_counts())

positions_gained
Gained Position    12809
Lost Position      11605
No Change           2345
Name: count, dtype: int64


**Note:**

The analysis shows that position gains occurred slightly more often than position losses. Out of **26,759 race results**, drivers gained positions in **12,809 cases**, while positions were lost in **11,605 cases**. In **2,345 cases**, drivers finished in the same position in which they started.

Overall, this indicates that gaining positions was slightly more common than losing positions during a race.

In [59]:
# Which drivers gained the most positions on average during races
driver_position_gain = (drivers_results_df.groupby(['surname', 'forename'])
                        .agg(avg_postions_gained=('positions_gained', 'mean'),
                             races=('positions_gained', 'count'))
                        )

In [67]:
driver_position_gain['races'].describe()

count    266.000000
mean      89.394737
std       72.675540
min       20.000000
25%       36.000000
50%       62.500000
75%      116.500000
max      404.000000
Name: races, dtype: float64

In [61]:
# Check for drivers with minimum of 20 races
driver_position_gain = (
    driver_position_gain[
        driver_position_gain['races'] >= 20
    ]
    .sort_values('avg_postions_gained', ascending=False)
)

# Check for result
print(driver_position_gain.head(20))

                         avg_postions_gained  races
surname      forename                              
Mazzacane    Gastón                 7.047619     21
Gené         Marc                   5.916667     36
Bellof       Stefan                 5.545455     22
Rothengatter Huub                   5.535714     28
Lamy         Pedro                  5.343750     32
Palmer       Jonathan               4.647727     88
Nakano       Shinji                 4.484848     33
Streiff      Philippe               4.481481     54
Beuttler     Mike                   4.448276     29
Badoer       Luca                   4.344828     58
Fittipaldi   Christian              4.000000     43
Rosier       Louis                  3.666667     39
Baumgartner  Zsolt                  3.450000     20
Takagi       Toranosuke             3.437500     32
Marques      Tarso                  3.384615     26
Monteiro     Tiago                  3.243243     37
Claes        Johnny                 3.200000     25
d'Ambrosio  